### PR TITLE
Added support for complex64 and complex128 that are named by IR.md.

### PR DESCRIPTION
**Pull request by donbox**
Added double for symmetry (given that we are supporting 64-bit doubles in complex128.

### CODE REVIEW

**Comment by donbox** [comment]
Note that the representation used for complex128 and complex64 is to simply pack the two float/double values next to one another in the raw_data field (little-endian, IEEE754, real followed by imaginary component).

Another equally valid mapping would be to simply use the float_data and double_data fields and encode using odd # elements for reals, even # elements for complex.  That mapping might be more convenient as it doesn't involve raw buffer manipulating.

File: onnx/onnx.proto [hunk 1]
```diff
@@ -195,14 +243,23 @@ message TensorProto {
...
   //
-  // When this raw_data field is used to store tensor value, it is expected
-  // to use the little-endian order. For floating data types, they should
-  // be of the IEEE 754 format. Boolean type is stored as bytes.
+  // When this raw_data field is used to store tensor value, elements MUST 
+  // be stored in as fixed-width, little-endian order. 
+  // Floating-point data types MUST be stored in IEEE 754 format. 
+  // Complex64 elements must be written as two consecutive FLOAT values, real component first.
+  // Complex128 elements must be written as two consecutive DOUBLE values, real component first.
+  // Boolean type MUST be written one byte per tensor element (00000001 for true, 00000000 for false).
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: BOOL should be stored in int32_data above (see comment there)

File: onnx/onnx.proto [hunk 2]
```diff
@@ -195,14 +243,23 @@ message TensorProto {
   // raw bytes field. The only exception is the string case, where one is
   // required to store the content in the repeated bytes string_data field.
   //
-  // When this raw_data field is used to store tensor value, it is expected
-  // to use the little-endian order. For floating data types, they should
-  // be of the IEEE 754 format. Boolean type is stored as bytes.
+  // When this raw_data field is used to store tensor value, elements MUST 
+  // be stored in as fixed-width, little-endian order. 
+  // Floating-point data types MUST be stored in IEEE 754 format.
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: for double it probably  makes sense to create double_data field separately

File: onnx/onnx.proto [hunk 3]
```diff
@@ -195,14 +243,23 @@ message TensorProto {
   // raw bytes field. The only exception is the string case, where one is
   // required to store the content in the repeated bytes string_data field.
   //
-  // When this raw_data field is used to store tensor value, it is expected
-  // to use the little-endian order. For floating data types, they should
-  // be of the IEEE 754 format. Boolean type is stored as bytes.
+  // When this raw_data field is used to store tensor value, elements MUST 
+  // be stored in as fixed-width, little-endian order. 
+  // Floating-point data types MUST be stored in IEEE 754 format. 
+  // Complex64 elements must be written as two consecutive FLOAT values, real component first.
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: why not use float_data directly and enable convention there for "real component first"

File: onnx/onnx.proto (hunk 2 above)
Comment (donbox) [reply onnx/onnx.proto]: Check the original PR. It's field 10.

File: onnx/onnx.proto (hunk 1 above)
Comment (donbox) [reply onnx/onnx.proto]: That is consistent with the spec and is explicitly called out as a supported DataType on int32_data. . 

The use of raw_data is optional but supported for all data types except for string. The comment text simply makes it unambiguous how things get encoded when the raw_data field is used.

File: onnx/onnx.proto (hunk 3 above)
Comment (donbox) [reply onnx/onnx.proto]: Thanks - I agree (see my comment at the top of this PR).

I updated the PR to support this.

**Comment by ezyang** [comment]
It is possible that IR.md wasn't supposed to mention complex64/complex128 as supported types. Which frameworks support these kinds of numbers? I know PyTorch (https://github.com/pytorch/pytorch/issues/755) and Caffe2 do not. Does CNTK? If no one supports it, it would be better to not put it into the IR (and just make sure we have enough extensibility points to add it later, if it becomes a thing.)

**Comment by nouiz** [comment]
In Theano, we have basic complexe support. But we are seeing an increasing
need for this. More models use that or mimic it with manual implementation.
So if this is already working in this PR, I would think it is good to
include it now.

**Comment by ezyang** [comment]
In the model of "tiered" support, we should be clear what expectations we have about frameworks supporting complex numbers, which is that, they are unlikely to be very good support by most backends, at least in the short (even mid) term.

**Comment by nouiz** [comment]
That seem a pretty good to me.

**Comment by yuanbyu** [comment]
@donbox the data types were initially in but removed at the end to keep things minimal for v1. Currently CNTK doesn't support complex64 and complex128, but we have started to see needs for them.

File: onnx/onnx.proto [hunk 4]
```diff
@@ -236,14 +264,32 @@ message TensorProto {
...
   optional bytes raw_data = 9;
+
+  // For double
+  // Complex64 tensors are encoded as a single array of doubles,
+  // with the real components appearing in odd numbered positions, 
+  // and the corresponding imaginary component apparing in the 
+  // subsequent even numbered position. (e.g., [1.0 + 2.0i, 3.0 + 4.0i]
+  // is encoded as [1.0, 2.0 ,3.0 ,4.0]  
+  // When this field is present, the data_type field MUST be DOUBLE or COMPLEX128
+  repeated double double_data = 10 [packed = true];
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: if you're changing this - could you please change a few places in onnx library? E.g. onnx/numpy_helper.py and a few others (just do the search)

File: onnx/onnx.proto [hunk 5]
```diff
@@ -68,12 +68,16 @@ message AttributeProto {
   optional bytes s = 4;               // UTF-8 string
   optional TensorProto t = 5;         // tensor value
   optional GraphProto g = 6;          // graph
-
+  optional uint64 ui = 12;            // uint
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: do we have concrete use cases for unsigned and high-precision numbers as attributes? I'm not opposed to them, but just curious whether we really need them at this point (few more parts for a switch and also less natural as now float-valued ones need to be looked up in both `f` and `d`).

**Comment by donbox** [comment]
@dzhulgakov - I've updated helper.py and numpy_helper.py to support the new data types.

**Pull request merged**